scripts/generate-method-action-types.ts

```diff
@@ -0,0 +1,536 @@
+  // Since we're just using the method reference in the handler type,
+  // we don't need the full signature - just return the method name
+  // The actual signature will be inferred from the controller class
+  return node.name ? (node.name as ts.Identifier).text : '';
+}
+
+/**
+ * Generates the action types file for a controller.
+ *
+ * @param controller - The controller information object.
+ */
+async function generateActionTypesFile(
+  controller: ControllerInfo,
+): Promise<void> {
+  const outputDir = path.dirname(controller.filePath);
+  const outputFile = path.join(
+    outputDir,
+    `${controller.name}-method-action-types.ts`,
+  );
+
+  const content = await generateActionTypesContent(controller, outputFile);
+  await fs.promises.writeFile(outputFile, content, 'utf8');
+}
+
+/**
+ * Generates the content for the action types file.
+ *
+ * @param controller - The controller information object.
+ * @param outputFile - The path to the output file.
+ * @returns The content for the action types file.
+ */
+async function generateActionTypesContent(
+  controller: ControllerInfo,
+  outputFile: string,
+): Promise<string> {
+  const controllerImportPath = `./${controller.name}`;
+
+  let content = `/**
+ * This file is auto generated by \`scripts/generate-method-action-types.ts\`.
+ * Do not edit manually.
+ */
+
+import type { ${controller.name} } from '${controllerImportPath}';
+
+`;
+
+  const actionTypeNames: string[] = [];
+
+  // Generate action types for each exposed method
+  for (const method of controller.methods) {
+    const actionTypeName = `${controller.name}${capitalize(method.name)}Action`;
+    const actionString = `${controller.name}:${method.name}`;
+
+    actionTypeNames.push(actionTypeName);
+
+    // Add the JSDoc if available
+    if (method.jsDoc) {
+      content += `${method.jsDoc}\n`;
+    }
+
+    content += `export type ${actionTypeName} = {
+  type: \`${actionString}\`;
+  handler: ${controller.name}['${method.name}'];
+};\n\n`;
+  }
+
+  // Generate union type of all action types
+  if (actionTypeNames.length > 0) {
+    const unionTypeName = `${controller.name}MethodActions`;
+    content += `/**
+ * Union of all ${controller.name} action types.
+ */
+export type ${unionTypeName} = ${actionTypeNames.join(' | ')};\n`;
+  }
+
+  // Lint the generated content
+  return await lintFileContent(`${content.trimEnd()}\n`, outputFile);
+}
+
+/**
+ * Capitalizes the first letter of a string.
+ *
+ * @param str - The string to capitalize.
+ * @returns The capitalized string.
+ */
+function capitalize(str: string): string {
+  return str.charAt(0).toUpperCase() + str.slice(1);
+}
+
+/**
+ * Lints TypeScript content using ESLint and returns the fixed content.
+ *
+ * @param content - The TypeScript content to lint.
+ * @param outputFilePath - The path where the file would be written (used for file type detection).
+ * @returns The linted content.
+ */
+async function lintFileContent(
```

Comment:
This is fine for now, but once we fix `yarn lint:fix` in https://github.com/MetaMask/core/pull/5204 then maybe we can put this in front of `yarn lint:eslint` in the `yarn lint`/`yarn lint:fix` command and then we don't have to do this.

Reply:
Actually, I wonder if we need to lint the file here at all?

If we go ahead and change `yarn lint` to:

```
yarn generate-method-action-types --check && yarn lint:eslint && echo && yarn lint:misc --check && yarn constraints && yarn lint:dependencies && yarn lint:teams
```

then the `yarn lint:eslint` step will do the linting for us. And if it tells us that file that we generated isn't compatible with our lint rules we can just fix this script.

How does that sound?